numpy/linalg/tests/test_linalg.py

```diff
@@ -1583,8 +1583,24 @@ def check_qr(self, a):
         assert_almost_equal(r2, r1)
 
     def test_qr_empty(self):
-        a = np.zeros((0, 2))
-        assert_raises(linalg.LinAlgError, linalg.qr, a)
+        for m, n in [(0, 0), (3, 0), (0, 3)]:
+            k = min(m, n)
+            a = np.empty((m, n))
+            a_type = type(a)
+            a_dtype = a.dtype
+
+            self.check_qr(a)
+
+            r = np.linalg.qr(a, mode='r')
+            assert_equal(r.dtype, a_dtype)
+            assert_(isinstance(r, a_type))
+            assert_equal(r.shape, (k, n))
```

Comment:
This shape test should probably be in `check_qr` anyway

Reply:
It's there implicitly.
```py
        # mode == 'reduced'
        q1, r1 = linalg.qr(a, mode='reduced')
        # ...
        assert_(r1.shape == (k, n))
        # ...

        # mode == 'r'
        r2 = linalg.qr(a, mode='r')
        # ...
        assert_almost_equal(r2, r1)
```
So I'm taking that bit out.

Reply:
Fair enough. I think the explicit shape check you had here was a little clearer, but it's not super important.